pmlc/dialect/eltwise/ops.td

```diff
@@ -191,47 +191,47 @@ def AsBoolOp : EltwiseOp<"as_bool", [NoSideEffect]> {
   }];
 }
 
-def AbsOp : UnaryOp<"abs">;
-def ACosOp : UnaryFloatOp<"acos">;
-def AddOp : BinaryOp<"add", [Commutative]>;
-def AndOp : BinaryIntOp<"and", [Commutative]>;
-def ASinOp : UnaryFloatOp<"asin">;
-def AssignOp : UnaryOp<"assign">;
-def ATanOp : UnaryFloatOp<"atan">;
-def CeilOp : UnaryOp<"ceil">;
-def CmpEqOp : CompareOp<"cmp_eq", [Commutative]>;
-def CmpGeOp : CompareOp<"cmp_ge">;
-def CmpGtOp : CompareOp<"cmp_gt">;
-def CmpLeOp : CompareOp<"cmp_le">;
-def CmpLtOp : CompareOp<"cmp_lt">;
-def CmpNeOp : CompareOp<"cmp_ne", [Commutative]>;
-def CosHOp : UnaryFloatOp<"cosh">;
-def CosOp : UnaryFloatOp<"cos">;
-def DivOp : BinaryOp<"div">;
-def ExpOp : UnaryFloatOp<"exp">;
-def FloorOp : UnaryOp<"floor">;
+def AbsOp : UnaryOp<"abs", [NoSideEffect]>;
+def ACosOp : UnaryFloatOp<"acos", [NoSideEffect]>;
+def AddOp : BinaryOp<"add", [Commutative, NoSideEffect]>;
+def AndOp : BinaryIntOp<"and", [Commutative, NoSideEffect]>;
+def ASinOp : UnaryFloatOp<"asin", [NoSideEffect]>;
+def AssignOp : UnaryOp<"assign", [NoSideEffect]>;
+def ATanOp : UnaryFloatOp<"atan", [NoSideEffect]>;
+def CeilOp : UnaryOp<"ceil", [NoSideEffect]>;
+def CmpEqOp : CompareOp<"cmp_eq", [Commutative, NoSideEffect]>;
+def CmpGeOp : CompareOp<"cmp_ge", [NoSideEffect]>;
+def CmpGtOp : CompareOp<"cmp_gt", [NoSideEffect]>;
+def CmpLeOp : CompareOp<"cmp_le", [NoSideEffect]>;
+def CmpLtOp : CompareOp<"cmp_lt", [NoSideEffect]>;
+def CmpNeOp : CompareOp<"cmp_ne", [Commutative, NoSideEffect]>;
+def CosHOp : UnaryFloatOp<"cosh", [NoSideEffect]>;
+def CosOp : UnaryFloatOp<"cos", [NoSideEffect]>;
+def DivOp : BinaryOp<"div", [NoSideEffect]>;
+def ExpOp : UnaryFloatOp<"exp", [NoSideEffect]>;
+def FloorOp : UnaryOp<"floor", [NoSideEffect]>;
 def IdentOp : UnaryOp<"ident">;
```

Comment:
`ident` appears to be necessary to deal with cases where otherwise we could wind up with invalid stripe code. There are 2 cases I know about:

1) A user specifies an output multiple times, but each output gets bound to a different buffer. This means we must copy the result from the original output to any ancillary ones.
2) A user specifies an output which refers directly to an input. This could cause an empty body in the pathological case (where the only output specified is also an input). The `ident` prevents the entire body from collapsing.

Reply:
We could rename this to: 
```
def CopyOp : UnaryOp<"copy">;
```

Reply:
👍